src/controllers/dynakube/activegate/reconciler.go

```diff
@@ -0,0 +1,139 @@
+package activegate
+
+import (
+	"context"
+	"fmt"
+
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1"
+	"github.com/Dynatrace/dynatrace-operator/src/dtclient"
+	"github.com/Dynatrace/dynatrace-operator/src/kubeobjects"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	AGSecretSuffix             = "-activegate-tenant-secret"
+	CommunicationEndpointsName = "communication-endpoints"
+	TenantTokenName            = "tenant-token"
+	TenantUuidName             = "tenant-uuid"
+)
+
+type Reconciler struct {
+	client.Client
+	apiReader client.Reader
+	instance  *dynatracev1beta1.DynaKube
+	scheme    *runtime.Scheme
+	apiToken  string
+	dtc       dtclient.Client
+}
+
+func NewReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, instance *dynatracev1beta1.DynaKube, apiToken string, dtc dtclient.Client) *Reconciler {
+	return &Reconciler{
+		Client:    clt,
+		apiReader: apiReader,
+		scheme:    scheme,
+		instance:  instance,
+		apiToken:  apiToken,
+		dtc:       dtc,
+	}
+}
+
+func (r *Reconciler) Reconcile() error {
+	if r.instance.Spec.ActiveGate.AGTenantSecret == "" {
+		err := r.reconcileActiveGateSecret()
+		if err != nil {
+			log.Error(err, "could not reconcile ActiveGate tenant secret")
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileActiveGateSecret() error {
+	agSecretData, err := r.getActiveGateTenantInfo()
+	if err != nil {
+		return fmt.Errorf("could not generate ag secret data: %w", err)
+	}
+
+	agSecret, err := r.createAGSecretIfNotExists(agSecretData)
+	if err != nil {
+		return fmt.Errorf("failed to create or update secret: %w", err)
+	}
+
+	return r.updateAGSecretIfOutdated(agSecret, agSecretData)
+}
+
+func (r *Reconciler) getActiveGateTenantInfo() (map[string][]byte, error) {
+	tenantInfo, err := r.dtc.GetActiveGateTenantInfo()
+
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return map[string][]byte{
+		TenantUuidName:             []byte(tenantInfo.UUID),
+		TenantTokenName:            []byte(tenantInfo.Token),
+		CommunicationEndpointsName: []byte(tenantInfo.Endpoints),
+	}, nil
+}
+
+func (r *Reconciler) createAGSecretIfNotExists(agSecretData map[string][]byte) (*corev1.Secret, error) {
+	var config corev1.Secret
+	err := r.apiReader.Get(context.TODO(), client.ObjectKey{Name: extendWithAGSecretSuffix(r.instance.Name), Namespace: r.instance.Namespace}, &config)
```

Comment:
nitpick: this line could use a few newlines or something to make it shorter :D

Reply:
Tip: I merged recently something to achieve that - `kubeobjects.Key()`
https://github.com/Dynatrace/dynatrace-operator/blob/master/src/kubeobjects/object.go#L5

Reply:
@toszr I just found out it's already implemented in `controller-runtime` as:
```
package client

// ObjectKeyFromObject returns the ObjectKey given a runtime.Object.
func ObjectKeyFromObject(obj Object) ObjectKey {
	return ObjectKey{Namespace: obj.GetNamespace(), Name: obj.GetName()}
}
```

Reply:
Hmm. I like `Key()` more than `ObjectKeyFromObject()` but thanks for finding it - I knew someone must have thought of it before.

Reply:
It can't be used here because name is calculated on the fly so it must be set explicitly.
Added newlines to make it more readable.